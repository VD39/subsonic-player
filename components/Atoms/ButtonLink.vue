<script setup lang="ts">
const props = withDefaults(
  defineProps<{
    disabled?: boolean;
    fullWidth?: boolean;
    icon?: Component | string;
    iconColor?: string;
    iconPosition?: IconPosition;
    iconSize?: number;
    iconWeight?: IconWeight;
    is?: ButtonLink;
    showText?: boolean;
    type?: string;
  }>(),
  {
    disabled: undefined,
    icon: undefined,
    iconColor: undefined,
    iconPosition: 'left',
    iconSize: 24,
    iconWeight: 'duotone',
    is: 'button',
    type: 'button',
  },
);

const type = props.is === 'button' ? props.type : undefined;
const isComponent =
  props.is === 'nuxt-link' ? resolveComponent('NuxtLink') : props.is;
</script>

<template>
  <component
    :is="isComponent"
    :class="[
      'centerItems',
      $style.buttonLink,
      {
        [$style.alignRight]: icon && iconPosition === 'right',
        [$style.disabled]: disabled,
        [`centerAll ${$style.fullWidth}`]: fullWidth,
      },
    ]"
    :disabled="disabled"
    :type="type"
    v-bind="$attrs"
  >
    <component
      :is="icon"
      v-if="icon"
      ref="iconComponent"
      :class="$style.icon"
      aria-hidden="true"
      :size="iconSize"
      :weight="iconWeight"
      :color="iconColor"
    />

    <span
      ref="text"
      :class="[
        'clamp',
        $style.text,
        {
          visuallyHidden: !showText,
        },
      ]"
    >
      <slot />
    </span>
  </component>
</template>

<style module>
.buttonLink {
  position: relative;
  gap: var(--default-space);
  padding: var(--default-space);
  white-space: nowrap;
}

.alignRight {
  flex-direction: row-reverse;
}

.fullWidth {
  width: var(--width-height-100);
}

.disabled {
  pointer-events: none;
  opacity: 0.25;
}

.icon {
  flex-shrink: 0;

  [opacity='0.2'] {
    opacity: 0.4;
    fill: var(--secondary-font-color);
  }
}

.text {
  display: block;
  margin-bottom: 1px;
}
</style>
