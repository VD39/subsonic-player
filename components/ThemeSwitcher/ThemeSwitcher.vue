<script setup lang="ts">
import IconButton from '@/components/Buttons/IconButton.vue';
import { useTheme } from '@/composables/useTheme';

const { isDarkTheme, toggleTheme } = useTheme();

const text = computed(() => {
  const mode = isDarkTheme.value ? 'light' : 'dark';
  return `Activate ${mode} mode`;
});
</script>

<template>
  <client-only>
    <IconButton
      :icon="isDarkTheme ? 'sun' : 'moon'"
      :title="text"
      @click="toggleTheme"
    >
      {{ text }}
    </IconButton>
  </client-only>
</template>
