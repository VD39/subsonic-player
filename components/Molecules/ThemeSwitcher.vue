<script setup lang="ts">
import ButtonLink from '@/components/Atoms/ButtonLink.vue';

const { isDarkTheme, toggleTheme } = useTheme();

const buttonProps = computed(() => ({
  icon: isDarkTheme.value ? ICONS.darkTheme : ICONS.lightTheme,
  text: `Activate ${isDarkTheme.value ? 'light' : 'dark'} mode`,
}));
</script>

<template>
  <client-only>
    <ButtonLink
      :icon="buttonProps.icon"
      :title="buttonProps.text"
      @click="toggleTheme"
    >
      {{ buttonProps.text }}
    </ButtonLink>
  </client-only>
</template>
