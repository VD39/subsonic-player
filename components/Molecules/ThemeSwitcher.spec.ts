import type { VueWrapper } from '@vue/test-utils';

import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import { mount } from '@vue/test-utils';

import ThemeSwitcher from './ThemeSwitcher.vue';

function factory(props = {}) {
  return mount(ThemeSwitcher, {
    props: {
      ...props,
    },
  });
}

describe('ThemeSwitcher', () => {
  let wrapper: VueWrapper;

  beforeEach(() => {
    wrapper = factory();
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('shows the correct icon prop', () => {
    expect(wrapper.findComponent(ButtonLink).props('icon')).toBe(
      ICONS.lightTheme,
    );
  });

  it('shows the correct title prop', () => {
    expect(wrapper.findComponent(ButtonLink).attributes('title')).toBe(
      'Activate dark mode',
    );
  });

  it('shows the correct text', () => {
    expect(wrapper.findComponent(ButtonLink).text()).toBe('Activate dark mode');
  });

  describe('when ButtonLink component emits a click event', () => {
    beforeAll(async () => {
      wrapper.findComponent(ButtonLink).vm.$emit('click');
      await wrapper.vm.$nextTick();
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('shows the correct icon prop', () => {
      expect(wrapper.findComponent(ButtonLink).props('icon')).toBe(
        ICONS.darkTheme,
      );
    });

    it('shows the correct title prop', () => {
      expect(wrapper.findComponent(ButtonLink).attributes('title')).toBe(
        'Activate light mode',
      );
    });

    it('shows the correct text', () => {
      expect(wrapper.findComponent(ButtonLink).text()).toBe(
        'Activate light mode',
      );
    });
  });
});
