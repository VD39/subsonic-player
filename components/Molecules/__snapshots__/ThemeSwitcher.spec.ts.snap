// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ThemeSwitcher > matches the snapshot 1`] = `
"<button class="centerItems buttonLink" type="button" title="Activate dark mode">
  <phmoon class="icon" aria-hidden="true" size="24" weight="duotone"></phmoon><span class="clamp text visuallyHidden">Activate dark mode</span>
</button>"
`;

exports[`ThemeSwitcher > when ButtonLink component emits a click event > matches the snapshot 1`] = `
"<button class="centerItems buttonLink" type="button" title="Activate light mode">
  <phsun class="icon" aria-hidden="true" size="24" weight="duotone"></phsun><span class="clamp text visuallyHidden">Activate light mode</span>
</button>"
`;
